sunpy/net/dataretriever/sources/goes.py

```diff
@@ -44,8 +59,105 @@ class XRSClient(GenericClient):
+        if matchdict["Time"].start < "2009-09-01" and matchdict["Time"].end >= "2009-09-01":
+            matchdict_before, matchdict_after = matchdict.copy(), matchdict.copy()
+            matchdict_before["Time"] = TimeRange(matchdict_before["Time"].start, '2009-08-31')
+            matchdict_after["Time"] = TimeRange('2009-09-01', matchdict_after["Time"].end)
+            metalist_before = self._get_metalist(matchdict_before)
+            metalist_after = self._get_metalist(matchdict_after)
+            metalist = metalist_before + metalist_after
+        else:
+            metalist = self._get_metalist(matchdict)
+
+        return QueryResponse(metalist, client=self)
+
+    def _get_metalist(self, matchdict):
+        """
+        Function to get the list of OrderDicts.
+        This makes it easier for when searching for overlapping providers.
+        """
+        metalist = []
+        # the data before the re-processed GOES 13, 14, 15 data.
+        if matchdict["Time"].end < "2009-09-01":
+            scraper = Scraper(self.baseurl_old, regex=True)
+            filemeta = scraper._extract_files_meta(
+                matchdict["Time"], extractor=self.pattern_old, matcher=matchdict)
+            for i in filemeta:
+                rowdict = self.post_search_hook(i, matchdict)
+                metalist.append(rowdict)
+
+        # if for some reason a user wants the old data.
+        elif (matchdict["Time"].end >= "2009-09-01" and matchdict["Provider"] == ["sdac"]):
+            scraper = Scraper(self.baseurl_old, regex=True)
+            filemeta = scraper._extract_files_meta(
+                matchdict['Time'], extractor=self.pattern_old, matcher=matchdict)
+            for i in filemeta:
+                rowdict = self.post_search_hook(i, matchdict)
+                metalist.append(rowdict)
+
+        # new data from NOAA.
+        else:
+            if matchdict["Time"].end >= "2017-02-07":
+                for sat in [16, 17]:
+                    formdict = {"SatelliteNumber": sat}
+                    urlpattern = self.baseurl_r.format(**formdict)
+                    scraper = Scraper(urlpattern)
+                    filemeta = scraper._extract_files_meta(
+                        matchdict["Time"], extractor=self.pattern_r, matcher=matchdict)
+                    for i in filemeta:
+                        rowdict = self.post_search_hook(i, matchdict)
+                        metalist.append(rowdict)
+
+            for sat in [13, 14, 15]:
+                formdict = {"SatelliteNumber": sat}
+                urlpattern = self.baseurl_new.format(**formdict)
+                scraper = Scraper(urlpattern)
+                filemeta = scraper._extract_files_meta(
+                    matchdict["Time"], extractor=self.pattern_new, matcher=matchdict)
+                for i in filemeta:
+                    rowdict = self.post_search_hook(i, matchdict)
+                    metalist.append(rowdict)
```

Comment:
rather than repeating the whole block here, could you not set a variable like `new_pattern = self.pattern_r` and the same for base url in the if block and then only write the actual scraping code once using those variables?

Reply:
hmm I'm not too sure if this would work due to the fact that there are days when they are available GOES 13, 14, 15, 16 and 17 data so needs to do this in an append method I think

Reply:
Ah yeah I think I see what you are saying. It still feels a little to repetitive here, I wonder if moving this pattern into a function which could be used in these 3 different conditions would be worth it?

Reply:
yeah thats probably the best idea, I'll do that